### PR TITLE
 Equality Generator now supports baseclasses

### PR DESCRIPTION
**Pull request by carldebilly**

### CODE REVIEW

File: src/Uno.CodeGen/EqualityGenerator.cs [hunk 1]
```diff
@@ -112,6 +123,15 @@ private void GenerateEquality(INamedTypeSymbol typeSymbol)
 					builder.AppendLineInvariant($"#warning {nameof(EqualityGenerator)}: you should add the partial modifier to the class {symbolNameWithGenerics}.");
 				}
 
+				if (baseTypeInfo.isBaseType && !baseTypeInfo.baseOverridesEquals)
+				{
+					builder.AppendLineInvariant($"#warning {nameof(EqualityGenerator)}: base type {typeSymbol.BaseType} does not override .Equals() method. It could lead to erronous results.");
```
Comment (jeromelaban) [inline src/Uno.CodeGen/EqualityGenerator.cs]: erroneous, or incorrect

File: doc/Equality Generation.md [hunk 2]
```diff
@@ -228,4 +229,20 @@ The generation logic for fields/properties in the class, using a first-match rul
 * Remember: **Computing _Hash Codes_ should be quick**. It is better to have a
```
Comment (jeromelaban) [inline doc/Equality Generation.md]: hash codes

File: doc/Equality Generation.md [hunk 3]
```diff
@@ -228,4 +229,20 @@ The generation logic for fields/properties in the class, using a first-match rul
...
   quick computed value than a completely unique value. Colision should be avoided,
   but never as a result of a big computation. That's why only the length is used
-  by default for collections.
\ No newline at end of file
+  by default for collections.
+* For best results, your class should be immutable. For this, you can use the
+  [GeneratedImmutable] attribute. There is an optional parameter on this attribute
+  to automatically generate equality.
```
Comment (jeromelaban) [inline doc/Equality Generation.md]: generate equality members

File: doc/Immutable Generation.md [hunk 4]
```diff
@@ -190,4 +192,66 @@ Important:
 * Complex properties \*\***MUST**\*\* be immutable entities.
   A complex property is when it's a defined type, not a CLR primitive.
 * Indexers are not supported.
-* _Events Properties_ are not supported.
\ No newline at end of file
+* _Events Properties_ are not supported.
+
+# FAQ
+
+## I need to use it with _Newtownsoft's JSON.NET_. How can I acheive that?
```
Comment (jeromelaban) [inline doc/Immutable Generation.md]: Add a link to the https://github.com/JamesNK/Newtonsoft.Json

File: doc/Immutable Generation.md [hunk 5]
```diff
@@ -190,4 +192,66 @@ Important:
...
+Example:
+``` csharp
+  MyEntity e = JsonConvert.DeserializeObject<MyEntity.Builder>(json);
+```
+
+## It's generating a lot of unused method. It's a waste.
+For most application the compiled code won't be significant. Assets
+in projects are usually a lot bigger than that.
+
+If you are using an _AOT_ linking tool, those unused methods will
```
Comment (jeromelaban) [inline doc/Immutable Generation.md]: linker tool (not aot) link: http://www.mono-project.com/docs/tools+libraries/tools/linker/ or https://github.com/dotnet/core/blob/master/samples/linker-instructions.md

File: doc/Immutable Generation.md [hunk 6]
```diff
@@ -190,4 +192,66 @@ Important:
...
+be removed for final binary.
+
+We think the cost of this unused code is cheaper than the potential
+bugs when writing and maintaining this code manually.
+
+## What is the usage of the `[Immutable]` and `[ImmutableBuilder]` attributes?
+* **ImmutableAttribute**:
+  The `[Immutable]` is used by other parts of _Uno_
```
Comment (jeromelaban) [inline doc/Immutable Generation.md]: legacy

File: doc/Immutable Generation.md [hunk 7]
```diff
@@ -190,4 +192,66 @@ Important:
...
+  (some are not published as opened source) to identify an entity has been
+  immutable.
+* **ImmutableBuilderAttribute**:
+  The `[ImmutableBuilder]` is used to designate the builder able to
```
Comment (jeromelaban) [inline doc/Immutable Generation.md]: is used to indicate which builder can be used to build the type marked as immutable

File: doc/Immutable Generation.md [hunk 8]
```diff
@@ -190,4 +192,66 @@ Important:
...
+of an immutable class is created, it's impossible to change it.
+(ok, it's possible by using reflection, but why would you do that?)
+
+But the builders are not thread safe. That means updating the same
+property of the same instance concurrently (from many threads) will
+produce unexpected result.
+
+## Can we reuse builders?
+Yes. You can continue to update the builder even after calling
+.ToImmutable()`. The created instance won't be updated.
```
Comment (jeromelaban) [inline doc/Immutable Generation.md]: Missing ` before .ToImmutable.

File: src/Uno.CodeGen/EqualityGenerator.cs (hunk 1 above)
Comment (carldebilly) [reply src/Uno.CodeGen/EqualityGenerator.cs]: Will be part of my next push

File: doc/Equality Generation.md (hunk 2 above)
Comment (carldebilly) [reply doc/Equality Generation.md]: Big change :-)  Will be part of my next push.

File: doc/Equality Generation.md (hunk 3 above)
Comment (carldebilly) [reply doc/Equality Generation.md]: Same.

File: doc/Immutable Generation.md (hunk 4 above)
Comment (carldebilly) [reply doc/Immutable Generation.md]: I'll put a link on the JSON's homepage instead ;-)

File: doc/Immutable Generation.md (hunk 5 above)
Comment (carldebilly) [reply doc/Immutable Generation.md]: Oooh I'll put those links.

File: doc/Immutable Generation.md (hunk 6 above)
Comment (carldebilly) [reply doc/Immutable Generation.md]: I don't agree it's a legacy attribute.

File: doc/Immutable Generation.md (hunk 7 above)
Comment (carldebilly) [reply doc/Immutable Generation.md]: I'll improve the text.

**Comment by carldebilly** [comment]
This will fix #3 .

File: doc/Equality Generation.md [hunk 9]
```diff
@@ -20,9 +22,7 @@
        public string C { get; } 
    }
    ```
-   
-   * Note: The `[GeneratedEquality]` attribute is implicit when you are
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply doc/Equality Generation.md]: Nope...  was removed after an internal review but the documentation wasn't updated. But I think we should bring back this feature.

**Pull request merged**